ts: update the recording of changefeed child metrics

jasonlmfong opened the pull request:
ts: record histogram child metrics

When child metrics collection was performed, it only picked up gauges and counters.
This commit adds the ability to record histograms as well.
This was done through exposing child histograms snapshots which were previously not accessible.

Epic: CRDB-55079
Release: None

--

ts: memoize child metric name transformation

Curently the child metric name transformations are being performed every polling cycle.
The memoization will ensure every subsequent poll does not need to allocate memory and cpu to transform the names.

Epic: CRDB-55079
Release: None

dhartunian reviewed (changes requested):
> @dhartunian reviewed 2 of 2 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jasonlmfong)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjX5TP78DqHUmXPVe_:-OfjX5TP78DqHUmXPVea:bi6sjp8):*
nit: commit message wrap.
___
*[`-- commits` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjXKd25APswb8pfmDd:-OfjXKd25APswb8pfmDe:b-i5pvt6):*
can you add more detail about why a bunch of histogram wrangling code is necessary? Seems like it's related to child histograms not exposing their internals (because they weren't needed before) and the manual work involved in calculating the percentiles, right?

nit: commit message wrap.
___
*[`-- commits` line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjYbIY7LecBtnGEhRs:-OfjYbIY7LecBtnGEhRt:b-tewbqe):*
nit: grammar. also consider rewording. "memoize child metric name transformation" might be clearer.
___
*[`-- commits` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjYrI22hWW1QCEsjaS:-OfjYrI22hWW1QCEsjaT:b-7ki0cn):*
Be clear about what motivated the introduction of the cache, and what you expect it to improve. This way someone can understand why this code is here in the future. Otherwise, it seems unmotivated on the surface especially since it's not linked to a ticket.
___
*[`pkg/server/status/recorder_test.go` line 1170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjXALZBCDoW0nX3GHT:-OfjXALZBCDoW0nX3GHU:b-309pqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d38fdd8625406c0837ed937018add5d504cd7a/pkg/server/status/recorder_test.go#L1170)):*
> ```Go
> 		{"5metrics_100children", 5, 100},
> 		{"10metrics_100children", 10, 100},
> 		{"10metrics_1024children", 10, 1024},
> ```

what is the purpose of doing different combos for benchmarks? shouldn't 1 benchmark scale to different number of child metrics?
___
*[`pkg/server/status/recorder.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_MnX8JFkZbLzahxT:-Ofj_MnX8JFkZbLzahxU:b-7aa50v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L197)):*
> ```Go
> 	// childMetricNameCache caches the encoded names for child metrics to avoid
> 	// rebuilding them on every recording. Uses sync.Map for lock-free reads.
> 	childMetricNameCache sync.Map
> ```

linter will tell you this, but you need to use `syncutil.Map` which supports generics.
___
*[`pkg/server/status/recorder.go` line 476 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_n6o42_GXBGeGoLQ:-Ofj_n6o42_GXBGeGoLR:b-3o8pof) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L476)):*
> ```Go
> 			timestampNanos: now.UnixNano(),
> 		}
> 		recorder.recordChangefeedChildMetrics(&data, &mr.childMetricNameCache)
> ```

did you consider putting the childMetricNameCache in the `registryRecorder` struct so you don't have to pass it around?
___
*[`pkg/server/status/recorder.go` line 960 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjZ1QSDA3GeaezMFRh:-OfjZ1QSDA3GeaezMFRi:be9arcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L960)):*
> ```Go
> func hashLabels(labels []*prometheusgo.LabelPair) uint64 {
> 	var h uint64 = 14695981039346656037 // FNV-1a offset basis
> 	const prime = 1099511628211
> ```

surprising to me that you have to do this manually. is there no util or stdlib thing that can accumulate a hash from a bunch of strings? you should use `hash/maphash` or `hash/fnv`.
___
*[`pkg/server/status/recorder.go` line 1122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_8i879eDGth454MH:-Ofj_8i99zFIHjD3mlXb:b28vj2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L1122)):*
> ```Go
> 			} else {
> 				metricName = prom.GetName(false /* useStaticLabels */) + metric.EncodeLabeledName(childMetric)
> 			}
> ```

can you extract the shared logic for the hash lookup? seems identical as the histogram one above.
___
*[`pkg/server/status/recorder_test.go` line 366 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjY3mt4CZvjjeq59vW:-OfjY3mt4CZvjjeq59vX:bai9ry2) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c39ba45d9ff0d16de15fa7c8ec10aa6405dd47/pkg/server/status/recorder_test.go#L366)):*
> ```Go
> 
> 		// Check for changefeed.stage.downstream_client_send.latency with scope labels
> 		if strings.Contains(ts.Name, "cr.node.changefeed.stage.downstream_client_send.latency") {
> ```

this is a weird construction for the test to conditionally evaluate so many varieties of cases. can this be encoded as a set of test case structs?
___
*[`pkg/server/status/recorder.go` line 1037 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjXghkCCPIJA4QIOuB:-OfjXghkCCPIJA4QIOuC:biptsxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c39ba45d9ff0d16de15fa7c8ec10aa6405dd47/pkg/server/status/recorder.go#L1037)):*
> ```Go
> 				}
> 				childMetricsCount++
> 			})
> ```

this code feels like it should be implemented by the child metric structs as a method, no? or is it specialized to this output. maybe this is just how we do non-child histogram recording?


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/status/recorder.go` line 476 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_n6o42_GXBGeGoLQ:-Ofjala85Dzvkby5vinT:b8fv1dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L476)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

did you consider putting the childMetricNameCache in the `registryRecorder` struct so you don't have to pass it around?
</blockquote></details>

the registryRecorder is being recreated when we call `GetTimeSeriesData()` so the cache doesn't live until the next iteration


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/status/recorder.go` line 1037 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjXghkCCPIJA4QIOuB:-Ofjla_s38DDbIijVUTM:bq93qgr) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c39ba45d9ff0d16de15fa7c8ec10aa6405dd47/pkg/server/status/recorder.go#L1037)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

this code feels like it should be implemented by the child metric structs as a method, no? or is it specialized to this output. maybe this is just how we do non-child histogram recording?
</blockquote></details>

we also do non-child histogram recordings in a similar fashion, it's under the `extractValue()` function


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
tftr, comments are addressed

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-:-OfzJW_Z3r_oyf7H3gI3:b-szzsho)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, and @dhartunian)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjX5TP78DqHUmXPVe_:-OfzJ9_G9n1Bs0RZv9Mn:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: commit message wrap.
</blockquote></details>

done
___
*[`-- commits` line 12 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjXKd25APswb8pfmDd:-OfzJAH_5w5QjrnO0SlO:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

can you add more detail about why a bunch of histogram wrangling code is necessary? Seems like it's related to child histograms not exposing their internals (because they weren't needed before) and the manual work involved in calculating the percentiles, right?

nit: commit message wrap.
</blockquote></details>

done
___
*[`-- commits` line 18 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjYbIY7LecBtnGEhRs:-OfzJAx2-VRTZKfklR0d:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

nit: grammar. also consider rewording. "memoize child metric name transformation" might be clearer.
</blockquote></details>

done
___
*[`-- commits` line 20 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjYrI22hWW1QCEsjaS:-OfzJBay1-_kzMuvZbrx:bb7f2mf):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Be clear about what motivated the introduction of the cache, and what you expect it to improve. This way someone can understand why this code is here in the future. Otherwise, it seems unmotivated on the surface especially since it's not linked to a ticket.
</blockquote></details>

done
___
*[`pkg/server/status/recorder.go` line 197 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_MnX8JFkZbLzahxT:-OfzJKPY26hEgdl6tKB-:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L197)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

linter will tell you this, but you need to use `syncutil.Map` which supports generics.
</blockquote></details>

done
___
*[`pkg/server/status/recorder.go` line 960 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjZ1QSDA3GeaezMFRh:-OfzJMPgC4WKH05mSFYq:b-npyl1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L960)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

surprising to me that you have to do this manually. is there no util or stdlib thing that can accumulate a hash from a bunch of strings? you should use `hash/maphash` or `hash/fnv`.
</blockquote></details>

done with hash/fnv
___
*[`pkg/server/status/recorder.go` line 1122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_8i879eDGth454MH:-OfzJPU_07Jg9QYzhZoX:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L1122)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

can you extract the shared logic for the hash lookup? seems identical as the histogram one above.
</blockquote></details>

done
___
*[`pkg/server/status/recorder_test.go` line 1170 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjXALZBCDoW0nX3GHT:-OfzJTRT2Dup9JXLQ58V:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/34d38fdd8625406c0837ed937018add5d504cd7a/pkg/server/status/recorder_test.go#L1170)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

what is the purpose of doing different combos for benchmarks? shouldn't 1 benchmark scale to different number of child metrics?
</blockquote></details>

done
___
*[`pkg/server/status/recorder_test.go` line 366 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfjY3mt4CZvjjeq59vW:-OfzJQ_e22TNBpqxoWNq:b-p0tnxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/94c39ba45d9ff0d16de15fa7c8ec10aa6405dd47/pkg/server/status/recorder_test.go#L366)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

this is a weird construction for the test to conditionally evaluate so many varieties of cases. can this be encoded as a set of test case structs?
</blockquote></details>

done in separate pr (commit 1)


<!-- Sent from Reviewable.io -->

jasonlmfong committed:
ts: record histogram child metrics

When child metrics collection was performed, it only picked up gauges and counters.
This commit adds the ability to record histograms as well.
This was done through exposing child histograms snapshots which were previously not accessible.

Epic: CRDB-55079
Release: None

dhartunian reviewed (commented):
Just a few more questions. Looks good otherwise.

> @dhartunian reviewed 4 of 4 files at r4, 1 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-:-OfzUMfx20TxUa7Pwvnf:b-itbxsp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, and @jasonlmfong)

___
*[`pkg/server/status/recorder.go` line 476 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_n6o42_GXBGeGoLQ:-OfzLgYsAFREwQy4NBmf:bgm01bh) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L476)):*
<details><summary><i>Previously, jasonlmfong (Jason Fong) wrote…</i></summary><blockquote>

the registryRecorder is being recreated when we call `GetTimeSeriesData()` so the cache doesn't live until the next iteration
</blockquote></details>

Sorry I was unclear. I was implying that you would pass in a pointer to it in the `registryRecorder` struct so that the methods can just read it without having to have an argument passed. Doesn't make a big diference.
___
*[`pkg/server/status/recorder.go` line 1006 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfzNiin9INWgHGxUKzW:-OfzNiin9INWgHGxUKzX:b-w2163m) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf6467e9836ee41a420db059168707187e1538e/pkg/server/status/recorder.go#L1006)):*
> ```Go
> 		}
> 
> 		recordMetric := func(name string, value float64) {
> ```

not sure this function definition is necessary since it's not capturing anything unique to the loop iteration.
___
*[`pkg/server/status/recorder.go` line 1064 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfzQKLG0ciY578MHRyF:-OfzQKLG0ciY578MHRyG:b-1y1l1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf6467e9836ee41a420db059168707187e1538e/pkg/server/status/recorder.go#L1064)):*
> ```Go
> 					var value float64
> 					if c.IsSummaryMetric {
> 						value = c.ComputedMetric(windowedSnapshot)
> ```

can you compute the corresponding snapshots in the if statements instead of doing both every time?


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
added a new section for hash collision detected as suggested by ai review

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-:-OfzaUre-A54mexp1KIO:bymerku)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @arjunmahishi, and @dhartunian)

___
*[`pkg/server/status/recorder.go` line 476 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-Ofj_n6o42_GXBGeGoLQ:-OfzW9G944K1LdMGFoU0:b-i0cyuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/8116a2ed402c92e224c13643e534094f7a715f72/pkg/server/status/recorder.go#L476)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Sorry I was unclear. I was implying that you would pass in a pointer to it in the `registryRecorder` struct so that the methods can just read it without having to have an argument passed. Doesn't make a big diference.
</blockquote></details>

Ah I see, it does make the interface cleaner
___
*[`pkg/server/status/recorder.go` line 1006 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfzNiin9INWgHGxUKzW:-Ofz_gTk4KX4LtMA7Yej:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf6467e9836ee41a420db059168707187e1538e/pkg/server/status/recorder.go#L1006)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

not sure this function definition is necessary since it's not capturing anything unique to the loop iteration.
</blockquote></details>

done
___
*[`pkg/server/status/recorder.go` line 1064 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-OfzQKLG0ciY578MHRyF:-Ofz_hAg2XtaPfoJ-jd1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6bf6467e9836ee41a420db059168707187e1538e/pkg/server/status/recorder.go#L1064)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

can you compute the corresponding snapshots in the if statements instead of doing both every time?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

jasonlmfong committed:
ts: memoize child metric name transformation

Curently the child metric name transformations are being performed every polling cycle.
The memoization will ensure every subsequent poll does not need to allocate memory and cpu to transform the names.

Epic: CRDB-55079
Release: None

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> @dhartunian reviewed 1 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158889#-:-Og33sqi1toR7veDoSOF:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, and @arjunmahishi)



<!-- Sent from Reviewable.io -->
